### ISSUE TITLE
Cross Origin pthread worker

### ISSUE THREAD

**Issue by msqr1** [emscripten-core/emscripten#21937]
I love the new inline worker change, but now we can't never use a 3rd-party CDN to serve those files again, because Worker construction requires the file to be in the same origin. Previously, we can just host the very small worker file on our server, while serving the heavy main file through a CDN, but now we can't anymore. Is there any way we could address this problem?

**Comment by sbc100** [comment]
Can you explain why this doesn't work?  Why cant you serve the main JS file from the CDN?

We do have  setting called `mainScriptUrlOrBlob` which controls where the main JS file finds itself.    I'm curious why this would be needed though since I would expect that loading JS file from the same location from which the main thread already loaded it would make the most sense and result in the best caching behaviour.  In fact, shouldn't the browser already have the main JS file cached, resulting in no extra JS fetches per-thread?

**Comment by msqr1** [comment]
I know it is cached, but it is still from cross origin. The thing that matter is the origin, not cached or not.

**Comment by sbc100** [comment]
Sorry I didn't read the title there.

I'm not too familiar with cross-origin limitations.   If you, or anyone else, has any ideas on how to make this work better that would be most welcome.

**Comment by sbc100** [comment]
So, to be clear, the security policy disassllows `new Worker` with a cross origin script but then allows the new worker to call `importScripts` from other origins?

**Comment by sbc100** [comment]
Trying to understand the issue, the only mention I see about same-origin restrictions in workers looks like it related to subworkers: 
 https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers.   Are you using subworkers?  Or is this some kind of additional non-default policy that you have setup?

**Comment by sbc100** [comment]
I tried this locally and the following `--pre-js` seemed to work:

```
if (!ENVIRONMENT_IS_PTHREAD) {
  Module['mainScriptUrlOrBlob'] = new Blob([`importScripts('${document.currentScript.src}');`])
}
```

**Comment by msqr1** [comment]
Did you try to serve that file through a cdn?

**Comment by sbc100** [comment]
Nope, I'm just going by the fact that `importScripts` from your CDN worked for you in the past.  So this is way to effectively revert to the prior behaviour.

**Comment by msqr1** [comment]
Yes, importscript in the worker seems to work fine. So now we create a blob worker that calls importscripts, which is allowed to be cross origin

**Comment by sbc100** [comment]
I'm confused by https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers though.   It seems to suggest that only sub-workers have this requirement by default?   So by default, using a CDN like this should work?

**Comment by sbc100** [comment]
Can you confirm that the `--pre-js` fix works for you your use case?

**Comment by msqr1** [comment]
I know it's weird, but see this: https://stackoverflow.com/questions/58098143/why-are-cross-origin-workers-blocked-and-why-is-the-workaround-ok

**Comment by sbc100** [comment]
So it sounds like maybe MDN is just wrong on this a not up-to-date with the spec, since it only mentions the same-origin requirement for sub-workers.

**Comment by sbc100** [comment]
Spec: https://html.spec.whatwg.org/multipage/webappapis.html#fetch-a-classic-worker-script

**Comment by msqr1** [comment]
Yeah it does

**Comment by msqr1** [comment]
> Can you confirm that the `--pre-js` fix works for you your use case?

It doesn't, I'm using modularity, so in the module callback document.currentScript is null. I used the _scriptName variable that keeps a reference to document.currentScript.src

**Comment by sbc100** [comment]
Can you share the full fix you came up with here?

**Comment by msqr1** [comment]
I took yours and
 ```
 if (!ENVIRONMENT_IS_PTHREAD) {
   Module['mainScriptUrlOrBlob'] = new Blob([`importScripts('${_scriptName}');`])
 }
 ```

**Comment by sbc100** [comment]
Does that work?  Looking at the code it seems that `_scriptName` is set only after the `--pre-js` code is included.

**Comment by sbc100** [comment]
(We could potentially move the `_scriptName` assignment earlier).

**Comment by msqr1** [comment]
Wait, I thought the _scriptName is declared before the iife is returned to be used in the function?

**Comment by sbc100** [comment]
> Wait, I thought the _scriptName is declared before the iife is returned to be used in the function?

I think MODULARIZE mode differs in this respect and injects `_scriptName` very early on.

**Comment by msqr1** [comment]
PR?

**Comment by sbc100** [comment]
I'd like to better understand the problem, and decide if we should do this default or make it a new option.   I also wonder if there is any downside to doing in the case when things are hosted on the origin.   For example, does it make debugging hard / more opaque when the URL passed to `new Workers` is a blob that then called `importScripts` vs just a URL?


For now, is it OK if you use your workaround?

**Comment by msqr1** [comment]
I'm happy with my fix, but I think this should be the default if we decide to go with it. People wont expect this bug when they switch to a new version that keeps them from using a CDN

**Comment by msqr1** [comment]
One more thing, if we decide to go with this, don't forget the Blob MIME type is js, otherwise it won't work on Firefox: https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker#:~:text=Strict%20MIME%20type%20checks%20for%20worker%20scripts

**Comment by sbc100** [comment]
@msqr1 I'm curious, are there some CDNs (e.g couldflare and fastly type things) that don't require you to do cross-origin stuff?  i.e. they take over your whole domain, so to speak?   I could be misunderstanding but I wonder how popular those two different types of solution are?

**Comment by msqr1** [comment]
There is no such CDN service that I know of does that. But I think Website hosting service does exactly what you say, they have servers everywhere that basically acts as CDN. Those servers serve your main page, your assets and everything else you have. You can basically treat it as your own domain while still utilizing their setup CDN. I don't think it would make sense for a CDN to take over your domain, because what CDN does is it help you minimize distance by utilizing the numerous servers available around the globe because you only have a few.

**Comment by msqr1** [comment]
CSP for worker-src must be set to allow blob for this to work

**Comment by msqr1** [comment]
> I also wonder if there is any downside to doing in the case when things are hosted on the origin. 

@sbc100, CSP needs to be set for worker-src or default-src to blob

**Comment by sbc100** [comment]
> > I also wonder if there is any downside to doing in the case when things are hosted on the origin.
> 
> @sbc100, CSP needs to be set for worker-src or default-src to blob for the workaround to spawn the worker. But at this point, why don't we just spawn the whole worker content straight from the blob already? Why importScript if we're using blob anyway?
> 
> In my opinion, if we choose to do this, I would make a new setting like CROSS_ORIGIN_PTHREAD that uses the new approach of spawning from the blob.

It seems reasonable to add a new option for this yes.   Feel tree to send a patch.   BTW, I would call the option PTHREAD_CROSS_ORIGIN so that is groups with the other PTHREAD_-prefixed settings.

I'm also curious to see what proportion of folks are affected by this.

**Comment by sbc100** [comment]
If  the workaround requires worker-src or default-src already then why not use the `importScript` method?  One advantage to this is debug-ability since the worker then knows the source of its source, as it were.  Also, I can imagine that the `importScripts` methods is much for friendly to the browser cache, and maybe even the v8 code case, and could be faster for that reason.

**Comment by msqr1** [comment]
You're right, I forgot about the cache

**Comment by DylanShang** [comment]
I have encountered the same problem, how is the progress now?

**Comment by DylanShang** [comment]
Another thing is  when I set ENVIRONMENT='web,webview,worker', `Module['mainScriptUrlOrBlob']` will not appear in the script, it will appear until I add `node` enviroment. Maybe we need to think about this case.

**Comment by msqr1** [comment]
I think you can try adding 'mainScriptUrlOrBlob' onto INCOMING_JS_MODULE_API. 
But I agree with you. We need an actual fix instead of a workaround.

**Comment by dr-matt** [comment]
We've been encountering this problem and forced to serve our assets from the main host for now rather than from a CDN. What is the current state of any fix in progress and/or best known workaround for `MODULARIZE`d components?

**Comment by msqr1** [comment]
https://github.com/emscripten-core/emscripten/issues/21937#issuecomment-2111187435

**Comment by mattantone** [comment]
@msqr1 thank you, unfortunately even with that `--pre-js` code you referenced, we still get errors like this (in chrome):
```
SecurityError: Failed to construct 'Worker': Script at 'https://my.cdn.com/app.407.16d08d17f64fc5e4ffec.chunk.js' cannot be accessed from origin 'https://my.host.com'
```
when the app tries to spawn a thread. Perhaps this is a different issue than the one discussed here, or perhaps we're missing something else?

**Comment by dr-matt** [comment]
@msqr1 thank you, unfortunately that `--pre-js` code you referenced still produces errors like this (in chrome):
```
SecurityError: Failed to construct 'Worker': Script at 'https://my.cdn.com/app.407.2c29c8e870e8747b722e.chunk.js' cannot be accessed from origin 'https://my.host.com'
```
when a thread is spawned by our component. Perhaps we're encountering a different issue than the one discussed here, or perhaps we're just missing something else?

**Comment by NolanK** [comment]
I was encountering this same issue and can confirm that the [pre-js code](https://github.com/emscripten-core/emscripten/issues/21937#issuecomment-2110699273) seems to do the trick.

**Comment by sbc100** [comment]
I'm going to see if I can write a unit test for this so we can be sure its fixed once and for all.

I wonder if there is some way to simulate this you multiple server all running on localhost?

**Comment by sbc100** [comment]
I have a fix up for review: https://github.com/emscripten-core/emscripten/pull/25581

**Comment by sbc100** [comment]
I just landed https://github.com/emscripten-core/emscripten/pull/2558 which adds a `-sCROSS_ORIGIN` setting.  Please let me how this works for you.

**Comment by dr-matt** [comment]
@sbc100 Brilliant, finally had a chance to try the `-sCROSS_ORIGIN` flag and it works perfectly. Cheers!

**Comment by sbc100** [comment]
Woohoo!
